File: exercises/max_water_container_11.java
public class max_water_container_11 {
    
}
 
// Original Solution
class Solution_11 {
    public int maxArea(int[] h) {
        int max = 0;
        int l = 0, r = h.length - 1;
        while (l < r) {
            if (h[l] < h[r]) {
                max = Math.max(max, h[l] * (r - l));
                l += 1;
            } else if (h[l] > h[r]) {
                max = Math.max(max, h[r] * (r - l));
                r -= 1;
            } else {
                max = Math.max(max, h[l] * (r - l));
                l += 1;
                r -= 1;
            }
        }
        return max;
    }
}
// More efficent solution using similar logic but skips iterations that are proven to be lesser than max
// class Solution {
//     public int maxArea(int[] height) {
//         int left = 0;
//         int right = height.length -1;
//         int max = 0;

//         while(left < right){
//             int minH = Math.min(height[left], height[right]);
//             int area = (right - left) * minH;
//             max = Math.max(max, area);
            
//             while(left < right && height[left] <= minH){
//                 left++;
//             }
//             while(left < right && height[right] <= minH){
//                 right--;
//             }
//         }
//         return max;
//     }
// }

File: exercises/max_water_container_11.py
from typing import List

# Original solution
class Solution:
    def maxArea(self, h: List[int]) -> int:
        n = 0
        m = len(h) - 1
        Max = 0
        while n < m:
            area = min(h[n], h[m]) * (m - n)
            if h[n] <= h[m]:
                n += 1
            else:
                m -= 1
            Max = max(Max, area)
        return Max 

# Faster Solution using the same idea
# class Solution:
#     def maxArea(self, h: List[int]) -> int:
#         area, l, r = 0, 0, len(h)-1
#         while l<r:
#             if h[l]<h[r]:
#                 area = max(area,h[l]*(r-l))
#                 l+=1
#             elif h[r]<h[l]:
#                 area = max(area,h[r]*(r-l))
#                 r-=1
#             else:
#                 area = max(area,h[r]*(r-l))
#                 r-=1
#                 l+=1
#         return area
